### DISCUSSION TITLE
How to support this?

### DISCUSSION THREAD

**Discussion by yisar** [webpack/webpack#16298] [Q&A]
```js
import fs from 'node:fs' // node
import path from 'npm:path' // deno
import manifest from 'vemix:manifest' // custom
```

How to match such files in webpack? They do not necessarily come from node_ Modules is a custom name, and then the content is written later after matching. This is easy to do in esbuild, but what about webpack? It will first report an error, saying that 'vemix: manifest' is not a valid format.

**Comment by alexander-akait** [comment]
What do you mean? You can use the `externals` option and provide empty module (or fallback) to them

**Reply by yisar** [reply]
If it is a resolve fallback, I need to customize the fallback.
```js
resolve: {
  fallback: {
    'vemix:manifest': `export default { // this is a custom virtual module, not a file path.
        aaa: bbb
    }`
  }
}
```

Here is a esbuild implement, I need implement the same feature in webpack.
https://github.com/yisar/venti/blob/main/packages/core/src/plugin.mjs#L55

**Reply by alexander-akait** [reply]
Please use `externals`, externals also are supporting the function notation, so you can create any things what you need

**Reply by Dante-dan** [reply]
I think `externals` used to be [**Prevent bundling**](https://webpack.js.org/configuration/externals/#externals) ,  `resolve` would be more appropriate.

Actually, it just want to custom a new Scheme.  [I have seen a way to resolve this by plugin](https://stackoverflow.com/questions/66278241/module-build-failed-unhandledschemeerror-reading-from-alias-path-is-not-han), but webpack only support `data:` or `file:` URIs by default.

**Comment by yisar** [comment]
@alexander-akait 
```js
function ({context, request}, callback) {
      console.log(request); // no no no not this, but the content of the file, not the request path !!!
      callback(null, request);
    }
```

So far, `externals` cannot meet the requirements. I don't need to replace the file `path`. I need to replace a customized `xxx: yyy` tag with a `string content`(real file content). If you can't understand, you can spend a minute reading the equivalent esbuild plugin.

**Reply by alexander-akait** [reply]
Sorry, you don't want to listen to me, we are not esbuild, we have loader/plugin conceptions (where you can implement an logic/transformation/minification/rewritting/etc).

And I ask you not to poke me into the code, I perfectly understand what esbuild do there, but you do not want to read the documentation.

If you need content of file you need to create a loader and change generate code there (but what is code, no information in the question), if you need to modify resolving logic - you have `resolver`, if you want to provide externals - you have `externals`, need fallback to another module - `resovler.fallback`, these things are made for different purposes, yes, in some places they can do the same things.

> How to match such files in webpack? They do not necessarily come from node_ Modules is a custom name, and then the content is written later after matching. This is easy to do in esbuild, but what about webpack? It will first report an error, saying that 'vemix: manifest' is not a valid format.

What should I understand from your question? You didn't specify what you want to do - provide content, do fallback, do another resolve, rewrite path, generate runtime? What?


```
function ({context, request}, callback) {
      console.log(request); // no no no not this, but the content of the file, not the request path !!!
      callback(null, request);
    }
```

Why not use `fs.readFile`, we provide all information to do any things. If you need plugin - you can register own schema and create a handler, but it will recreate more code than `externals`.

Also you can create `mock/vemix_manifest.js` and specify it in `externals`.

There are a lot of possible solutions for your problem, If you want a full answer, then please provide a real use case and a problem - what will be in `vemix:` (and other schemas), what we should do with `node:fs` (provide fs or do mock), what is your target (web/node/deno/etc).
